packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -95,10 +143,47 @@ class RenderParagraph extends RenderBox {
     span?.recognizer?.addPointer(event);
   }
 
+  bool _hasVisualOverflow = false;
+  TextPainter _overflowPainter;
+  ui.Shader _overflowShader;
+
   @override
   void performLayout() {
     _layoutText(constraints);
     size = constraints.constrain(_textPainter.size);
+
+    final bool didOverflowWidth = size.width < _textPainter.width;
+    // TODO(abarth): We're only measuring the sizes of the line boxes here. If
+    // the glyphs draw outside the line boxes, we might thing that there isn't
+    // visual overflow when there actually is visual overflow.
```

Comment:
i'm more concerned about the opposite case, though, where we have slight horizontal overflow, start clipping, and as a side-effect clip the intentional vertical visual overflow.